### DISCUSSION TITLE
Question about tmodel(t) set in GAMS code.

### DISCUSSION THREAD

**Discussion by yangqiu91** [NREL/ReEDS-2.0#58] [Q&A: Features] [answered]
Hi ReEDS team, 

I am trying to understand the GAMS code. However, I am confused by the tmodel(t) set in GAMS code.

In the b_inputs.gms file, tmodel(t) is defined as a set, but it has no value (because tmodel(t) is set as “no”, see following screenshot). I also looked at the generated inputs.gdx file, and the tmodel(t) is indeed an empty set. In the c_supplymodel.gms file, tmodel(t) is used as a logical condition to declare many equations. I am wondering if tmodel(t) is an empty set, how can it be used as a logical condition to declare equations? Because if one logical condition is an empty set, it can lead to empty domain for the equation. Is that right? 
![image](https://user-images.githubusercontent.com/36561597/133195856-a64e1e6a-eea4-4982-b19e-cc188d097e36.png)

I guess this might be a simple trick, but I just started to code with GAMS, so I would really appreciate it if anyone can provide an explanation on this. Thanks.

**Comment by boxblox** [comment, marked as answer]
hi there, `tmodel` is a dynamic set and therefore it can be assigned at any point before the solve statement and all the equations that will get sent to the solver (which are included in the `model` declaration) will reflect the new assignment.  you can see this in `d_solveroneyear.gms` on line 58 and 59... just before the solve statement a year `%cur_year%` is added to the set `tmodel`.  hope that helps. --adam

**Reply by yangqiu91** [reply]
Hi Adam,

Thank you very much for the reply. That makes a lot of sense. 
In that case, under the seq solve structure. `tmodel(t)` would be a set with only 1 value (the current year) in each solve, and it will change as the `%cur_year%` changes. Right?

**Reply by boxblox** [reply]
i like to answer questions about gams and the syntax (because I work at gams)... but I am not an expert at the NREL model.  I don't know enough about the larger seq solve structure to say either way.  

However, I will say that if you want to solve only one year at a time it might be worth creating `tmodel` as a `singleton set` rather than `set`.  you can create a `singleton set` over a domain like `singleton set tmodel(t)`... and then within some sort of loop you can keep setting `tmodel(t) = yes;` and it will always reset all elements to `no` except for the current `t`.  If `tmodel` is just a set, then you will want to make sure that at some point before the next solve the previous `yes` elements are reset to `no`.  Otherwise you could end up sending multiple years worth of equations to the solver.

**Reply by yangqiu91** [reply]
Make sense. Thank you again!

**Reply by wesleyjcole** [reply]
When running ReEDS sequentially, the `tmodel ` set only includes one year so that the model will only solve one year at a time.  However, when ReEDS is set to run intertemporally or using a sliding window, `tmodel` includes all the years that are to be solved.  In this way the equations can be identical regardless of the number of years to be solved.
